### PR TITLE
LinuxProcessList: use openat instead of building path strings

### PR DESCRIPTION
**Pull request by cgzones**
openat() is available since Linux 2.6.16

### CODE REVIEW

**Comment by BenBE** [comment]
`readlinkat` and `fstatat` have been available since then too and yet we put them in `Compat.h` …

File: linux/LinuxProcessList.c [hunk 1]
```diff
@@ -1151,12 +1150,18 @@ static bool LinuxProcessList_recurseProcTree(LinuxProcessList* this, const char*
...
 
       // Exception handler.
 
 errorReadingProcess:
       {
+         if (procFd >= 0)
+            close(procFd);
+
          if (preExisting) {
             ProcessList_remove(pl, proc);
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1166]: Shouldn't this trigger the diff highlighting logic?

File: linux/LinuxProcessList.c [hunk 2]
```diff
@@ -1222,49 +1227,65 @@ static inline void LinuxProcessList_scanMemoryInfo(ProcessList* this) {
...
       unsigned long long int compr_data_size = 0;
 
       if (!fscanf(disksize_file, "%llu\n", &size) ||
           !fscanf(mm_stat_file, "    %llu       %llu", &orig_data_size, &compr_data_size)) {
-         fclose(disksize_file);
          fclose(mm_stat_file);
-         break;
+         fclose(disksize_file);
+         close(zramFd);
+         continue;
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1274]: Why not just force `size`, `orig_data_size` and `compr_data_size` all to zero here and fall through?

Or invert this condition and handle the three additions from below in here.

File: linux/LinuxProcessList.c [hunk 3]
```diff
@@ -1605,5 +1626,13 @@ void ProcessList_goThroughEntries(ProcessList* super, bool pauseProcessUpdate) {
...
-   LinuxProcessList_recurseProcTree(this, PROCDIR, NULL, period, tv);
+   unsigned long long now = tv.tv_sec * 1000ULL + tv.tv_usec / 1000ULL;
+
+   int procDirfd = open(PROCDIR, O_RDONLY | O_DIRECTORY | O_NOFOLLOW);
+   if (procDirfd < 0) {
+      CRT_fatalError("Can not open the root process directory " PROCDIR);
+   }
+
+   /* LinuxProcessList_recurseProcTree() closes procDirfd */
+   LinuxProcessList_recurseProcTree(this, procDirfd, NULL, period, now);
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1637]: Not quite happy with this design. IMHO a resource provided by the caller should be freed by the caller. Makes for a few more LOC, but is cleaner from a resource management PoV IMO.

File: linux/LinuxProcessList.c (hunk 1 above)
Comment (cgzones) [reply linux/LinuxProcessList.c:1188]: Diff highlighting is fixed now

File: linux/LinuxProcessList.c (hunk 2 above)
Comment (cgzones) [reply linux/LinuxProcessList.c]: dropped the zram changes, no real gain

File: Action.c [hunk 4]
```diff
@@ -216,6 +216,16 @@ static Htop_Reaction actionToggleKernelThreads(State* st) {
...
 static Htop_Reaction actionToggleUserlandThreads(State* st) {
    st->settings->hideUserlandThreads = !st->settings->hideUserlandThreads;
+
+   if (st->settings->hideUserlandThreads) {
+      /* Do not show userland threads as tombed */
+      ProcessList_noTombedUserlandThreads(st->pl);
+   } else {
+      /* Do not show userland threads as new */
+      ProcessList_scan(st->pl, false);
+      ProcessList_noNewUserlandThreads(st->pl);
```
Comment (BenBE) [inline Action.c:226]: This discards legitimately new userland threads.

File: configure.ac [hunk 5]
```diff
@@ -88,7 +88,7 @@ AC_TYPE_UID_T
 # ----------------------------------------------------------------------
 AC_FUNC_CLOSEDIR_VOID
 AC_FUNC_STAT
-AC_CHECK_FUNCS([fstatat memmove readlinkat strdup strncasecmp strstr])
+AC_CHECK_FUNCS([fstatat memmove openat readlinkat strdup strncasecmp strstr])
```
Comment (BenBE) [inline configure.ac:91]: What about splitting this list one function per line?

File: Action.c (hunk 4 above)
Comment (cgzones) [reply Action.c:226]: With this patch userland threads are not parsed at all when disabled. So the are legitimately new on first scan after toggle-ling the setting.
It will only affect userland threads started **exactly** at the same time as the setting is switched, as there is one scan before (line 225), and one scan directly after https://github.com/htop-dev/htop/blob/be568b1153206a067fd09c0f7d2dcea3d7dbaa68/ScreenManager.c#L115 (cause we return `HTOP_RECALCULATE`)

Comment (BenBE) [reply Action.c:226]: What about tracking their PIDs (with `show==false`), but not filling in any other details while hiding userland threads?

Comment (cgzones) [reply Action.c]: moved the thread skipping to #334

File: linux/LinuxProcessList.c (hunk 3 above)
Comment (cgzones) [reply linux/LinuxProcessList.c]: reworked

File: configure.ac [hunk 6]
```diff
@@ -88,7 +88,7 @@ AC_TYPE_UID_T
 # ----------------------------------------------------------------------
 AC_FUNC_CLOSEDIR_VOID
 AC_FUNC_STAT
-AC_CHECK_FUNCS([faccessat fstatat memmove readlinkat strdup strncasecmp strstr])
+AC_CHECK_FUNCS([faccessat fstatat openat readlinkat])
```
Comment (BenBE) [inline configure.ac:91]:
```suggestion
AC_CHECK_FUNCS([\
   faccessat\
   fstatat\
   openat\
   readlinkat\
])
```

Comment (cgzones) [reply configure.ac]: changed

File: configure.ac (hunk 5 above)
Comment (cgzones) [reply configure.ac]: done

**Comment by BenBE** [comment]
Can you summarize what is missing for this PR?

**Comment by cgzones** [comment]
Tested (normal and compat version).

*Should* work.

**Review by natoscott — APPROVED** [review]
LGTM.

File: Compat.h [hunk 7]
```diff
@@ -22,6 +26,34 @@ int Compat_fstatat(int dirfd,
...
+
+typedef int openat_arg_t;
+
+static inline void Compat_openatArgClose(openat_arg_t dirfd) {
+   close(dirfd);
+}
+
+static inline int Compat_openat(openat_arg_t dirfd,
+                                const char* pathname,
+                                int flags) {
```
Comment (BenBE) [inline Compat.h:39]: No need to wrap this argument list …

File: Compat.h [hunk 8]
```diff
@@ -22,6 +26,34 @@ int Compat_fstatat(int dirfd,
...
+
+typedef const char* openat_arg_t;
+
+static inline void Compat_openatArgClose(openat_arg_t dirpath) {
+   (void)dirpath;
+}
+
+int Compat_openat(openat_arg_t dirpath,
+                  const char* pathname,
+                  int flags);
```
Comment (BenBE) [inline Compat.h:53]: Dito.

File: linux/LinuxProcessList.c [hunk 9]
```diff
@@ -1193,9 +1175,16 @@ static bool LinuxProcessList_readCmdlineFile(Process* process, const char* dirna
...
    /* execve could change /proc/[pid]/exe, so procExe should be updated */
-   xSnprintf(command, sizeof(command), "%s/%s/exe", dirname, name);
-   if ((amtRead = readlink(command, filename, sizeof(filename) - 1)) > 0) {
+#if defined(HAVE_READLINKAT) && defined(HAVE_OPENAT)
+   if ((amtRead = readlinkat(procFd, "exe", filename, sizeof(filename) - 1)) > 0) {
+#else
+   char path[4096];
+   xSnprintf(path, sizeof(path), "%s/exe", procFd);
+   if ((amtRead = readlink(path, filename, sizeof(filename) - 1)) > 0) {
+#endif
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1187]: I'd prefer if we could avoid starting a block within conditionally compiled sections of code.
```suggestion
#if defined(HAVE_READLINKAT) && defined(HAVE_OPENAT)
   amtRead = readlinkat(procFd, "exe", filename, sizeof(filename) - 1);
#else
   char path[4096];
   xSnprintf(path, sizeof(path), "%s/exe", procFd);
   amtRead = readlink(path, filename, sizeof(filename) - 1);
#endif
   if (amtRead > 0) {
```

File: linux/LinuxProcessList.c [hunk 10]
```diff
@@ -1484,12 +1492,20 @@ static bool LinuxProcessList_recurseProcTree(LinuxProcessList* this, const char*
...
       proc->updated = true;
+
+      Compat_openatArgClose(procFd);
+
       continue;
 
       // Exception handler.
 
 errorReadingProcess:
       {
```
Comment (BenBE) [inline linux/LinuxProcessList.c:1503]: Any reason for this extra code block for grouping? Doesn't it do without?

**Review by BenBE — COMMENTED** [review]
Minor stylistic notes. Apart from this LGTM.
